text/3845-repr-ordered-fields.md

```diff
@@ -0,0 +1,318 @@
+
+```
+warning: use of `repr(C)` in type `Foo`
+  --> src/main.rs:14:10
+   |
+14 |     #[repr(C)]
+   |       ^^^^^^^ help: consider switching to `repr(ordered_fields)`
+   |     struct Foo {
+   |
+   = note: `#[warn(edition_2024_repr_c)]` on by default
+   = note: `repr(C)` is planned to change meaning in the next edition to match the target platform's layout algorithm. This may change the layout of this type on certain platforms. To keep the current layout, switch to `repr(ordered_fields)`
+```
+
+After enough time has passed, and the community has switched over:
+This makes it easier to tell *why* the `repr` was applied to a given struct. If `repr(C)`, it's about FFI and interop. If `repr(ordered_fields)`, then it's for a dependable layout.
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `repr(C)`
+
+> The `C` representation is designed for one purpose: creating types that are interoperable with the C Language.
+> 
+> This representation can be applied to structs, unions, and enums. The exception is [zero-variant enums](https://doc.rust-lang.org/stable/reference/items/enumerations.html#zero-variant-enums) for which the `C` representation is an error.
+> 
+> - edited version of the [reference](https://doc.rust-lang.org/stable/reference/type-layout.html#the-c-representation) on `repr(C)`
+
+The exact algorithm is deferred to whatever the default target C compiler does with default settings (or if applicable, the most commonly used settings). 
+## `repr(ordered_fields)` 
+
+> The `ordered_fields` representation is designed for one purpose: create types that you can soundly perform operations on that rely on data layout such as reinterpreting values as a different type
+> 
+> This representation can be applied to structs, unions, and enums.
+> 
+> - edited version of the [reference](https://doc.rust-lang.org/stable/reference/type-layout.html#the-c-representation) on `repr(C)`
+### struct
+Structs are laid out in memory in declaration order, with padding bytes added as necessary to preserve alignment.
+And their alignment would be the same as their most aligned field.
+
+```rust
+// size 16, align 4
+#[repr(ordered_fields)]
+struct FooStruct {
+    a: u8,
+    b: u32,
+    c: u16,
+    d: u32,
+}
+```
+
+Would be laid out in memory like so
+
+```
+a...bbbbcc..dddd
+```
+### union
+Unions would be laid out with the same size as their largest field, and the same alignment as their most aligned field. 
+
+```rust
+// size 4, align 4
+#[repr(ordered_fields)]
+union FooUnion {
+    a: u8,
+    b: u32,
+    c: u16,
+    d: u32,
+}
+```
+
+Would have the same layout as `u32`.
+### enum
+The enum discriminant will be the smallest signed integer type which can hold all of the discriminant values (unless otherwise specified). The discriminants are assigned such that each variant without an explicit discriminant is exactly one more than the previous variant in declaration order. 
```

Comment:
I don't think this matches how enum discriminants work for `repr(C)`, but I'm not sure what algorithm is used for `repr(C)`. If someone knows, please LMK, and I'll update this section accordingly.

Reply:
I think I will change this to just say that it will use the same algorithm as `repr(C)` (so that the two reprs actually do match), and add a lint to ensure that `repr(ordered_fields)` is used with `repr(uN)`/`repr(iN)` (the lint can suggest which integer to use).